### ISSUE TITLE
Cannot apply cart lines change while useBuyerJourneyIntercept is 'block'

### ISSUE THREAD

**Issue by aurelienbobenrieth** [Shopify/ui-extensions#761]
## Please list the package(s) involved in the issue, and include the version you are using

`@shopify/checkout-ui-extensions-react` v0.24.0

## Describe the bug

I am building a custom app for a client of mine. Here is the use case:

The site is a marketplace in which the site is taking a commission on each sale.

We are usign a custom app + checkout ui extension to compute the commission based on the cart content:
1. Customer arrives on the checkout
2. CartLines are sent to our app backend
3. App backend creates and publishes the commission (created as a digital product) and returns the product JSON
4. Checkout UI extension gets the JSON as a response and applies it to the cart using `useApplyCartLinesChange()`

Of course I do not want the customers to be able to go any further in their checkout flow during the computing process.

However when blocking the buyer journey `useApplyCartLinesChange` fails with a generic error (that made it so hard to debug):
`{type: 'error', message: 'Could not apply line item changes'}`.

## Steps to reproduce the behavior:

To reproduce the issue, you can implement something like this:

```typescript
enum FetchingState {
  Waiting = 'waiting',
  ShouldLoad = 'should_load',
  Loading = 'loading',
  Loaded = 'loaded',
  Error = 'error',
}

// ...
const [fetchingState, setFetchingState] = useState<FetchingState>(
  FetchingState.Waiting
);
const applyCartLinesChange = useApplyCartLinesChange();
const canBlockProgress = useExtensionCapability("block_progress");

// ...

useEffect(() => {
  const createAndAddCommissionProductToCart = async () => {
    const application = await applyCartLinesChange({
      type: "addCartLine",
      merchandiseId: "YOUR_VARIANT_ID_HERE",
      quantity: 1,
    });
  };

  if (fetchingState === FetchingState.ShouldLoad) {
    setFetchingState(FetchingState.Loading);

    // This will fail since the buyer journey === 'block'
    createAndAddCommissionProductToCart();
  }
}, [fetchingState]);

// ...

if (canBlockProgress) {
  useBuyerJourneyIntercept(() => {
    /**
     * Should be:
     * if (fetchingState !== FetchingState.Loaded)
     * but currently impossible because of the issue I am describing
     */
    if (fetchingState === FetchingState.Error) {
      return {
        behavior: 'block',
        reason: InvalidResultReason.CommissionFetchingIssue,
        perform: () => {
          setShowBlockingBanner(true);
        },
      };
    }

    return {
      behavior: 'allow',
      perform: () => {
        setShowBlockingBanner(false);
      },
    };
  });
}
```

## Expected behavior

Basically I would like to be able to block buyer journey and still be able to process some cart updates behind the scenes.

At the moment my use case described above is impossible, and if a customer goes fast enough to proceed his order, he might be able to complete his order whithout being charged a commission.

**Comment by kumar303** [comment]
Hi @aurelienbobenrieth. Thanks for taking the time to report a bug. Could you provide a concise, runnable version of extension code that reproduces the bug?

I can't run the code above because it's missing some variables. Also, it looks like you commented out the code that actually triggers the bug?

>     /**
>     * Should be:
>     * if (fetchingState !== FetchingState.Loaded)
>     * but currently impossible because of the issue I am describing
>     */

I verified that simple blocking is working in production with this extension (below). There could be a timing issue that I don't fully understand, though.

<details>
<summary>Example code</summary>

```tsx
import { useState } from "react";
import {
  render,
  Banner,
  useExtensionCapability,
  useBuyerJourneyIntercept,
} from "@shopify/checkout-ui-extensions-react";

render("Checkout::Dynamic::Render", () => <App />);

function App() {
  const [bannerText, setBannerText] = useState<null | string>(null);
  const canBlockProgress = useExtensionCapability("block_progress");

  useBuyerJourneyIntercept(() => {
    if (!canBlockProgress) {
      setBannerText(
        "This extension is not allowed to block checkout progress. Make sure `block_progress = true` in [capabilities]"
      );
      return { behavior: "allow" };
    }
    return {
      behavior: "block",
      reason: "example reason",
      perform: () => {
        setBannerText("An extension is blocking checkout progress");
      },
    };
  });

  if (bannerText) {
    return <Banner status="critical">{bannerText}</Banner>;
  }

  return <Banner>The extension is not blocking checkout</Banner>;
}
```

</details>

**Comment by aurelienbobenrieth** [comment]
@misha-reyzlin-shopify Great news! Thank you very much!
